lib/spack/spack/binary_distribution.py

```diff
@@ -631,3 +708,114 @@ def get_keys(install=False, trust=False, force=False):
                 else:
                     tty.msg('Will not add this key to trusted keys.'
                             'Use -t to install all downloaded keys')
+
+
+def read_from_url(file_uri):
+    file_contents = None
+
+    try:
+        url = urlopen(file_uri)
+    except URLError as url_err:
+        msg = 'Unable to open url {0} due to {1}'.format(
+            file_uri, url_err.message)
+        raise spack.error.SpackError(msg)
+    except Exception as expn:
+        msg = 'Error getting file contents at {0} due to {1}'.format(
+            file_uri, expn.message)
+        raise spack.error.SpackError(msg)
+
+    file_contents = url.read()
+
+    if not file_contents:
+        msg = 'Unable to read file contents at {0}'.format(file_uri)
+        raise spack.error.SpackError(msg)
+
+    return file_contents
+
+
+def needs_rebuild(spec, mirror_url, buildcache_index):
+    pkg_name = spec.name
+    pkg_version = spec.version
+
+    tty.msg('Checking {0}-{1}'.format(pkg_name, pkg_version))
+
+    spec.concretize()
+    pkg_hash = spec.dag_hash()
+    pkg_full_hash = spec.full_hash()
+
+    rebuild_spec = {
+        'short_spec': spec.short_spec,
+        'hash': pkg_hash
+    }
+
+    if buildcache_index:
+        # just look in the index we already fetched
+        if pkg_hash in buildcache_index:
+            # At least remote binary mirror knows about it, so if the
+            # full_hash doesn't match (or remote end doesn't know about
+            # the full_hash), then we trigger a rebuild.
+            remote_pkg_info = buildcache_index[pkg_hash]
+            if ('full_hash' not in remote_pkg_info or
```

Comment:
this logic LGTM, but can you add a note that eventually it'll be less complicated (when the dag_hash is the full_hash)?